File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    void nse(vector<int> arr, vector<int>& nse) {
        int n = arr.size();
        stack<int> st;
        for(int i=n-1; i>=0; i--) {
            while(!st.empty() && arr[st.top()] >= arr[i]) st.pop();
            nse[i] = st.empty() ? n : st.top();
            st.push(i); 
        }
    }
    void pse(vector<int> arr, vector<int>& pse) {
        int n = arr.size();
        stack<int> st;
        for(int i=0; i<n; i++) {
            while(!st.empty() && arr[st.top()] >= arr[i]) st.pop();
            pse[i] = st.empty() ? -1 : st.top();
            st.push(i); 
        }
    }

    int largestRectangleArea(vector<int>& h) {
        int ma = 0;
        int n = h.size();
        vector<int> nsee(n);
        vector<int> psee(n);
        nse(h, nsee);
        pse(h, psee);
        for (int i = 0; i < n; i++) {
            ma = max(ma, h[i]*(nsee[i]-psee[i]-1));
        }
        return ma;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
import java.util.*;

class Solution {

    public int largestRectangleArea(int[] heights) {

        int n = heights.length;

        int[] pse = new int[n];
        int[] nse = new int[n];

        Stack<Integer> stack = new Stack<>();

        // PSE
        for (int i = 0; i < n; i++) {

            while (!stack.isEmpty() && heights[stack.peek()] >= heights[i]) {
                stack.pop();
            }

            if (stack.isEmpty()) {
                pse[i] = -1;
            } else {
                pse[i] = stack.peek();
            }

            stack.push(i);
        }

        // Empty the stack
        stack.clear();

        // NSE
        for (int i = n - 1; i >= 0; i--) {

            while (!stack.isEmpty() && heights[stack.peek()] >= heights[i]) {
                stack.pop();
            }

            if (stack.isEmpty()) {
                nse[i] = n;
            } else {
                nse[i] = stack.peek();
            }

            stack.push(i);
        }

        // Calculate area
        int ans = 0;

        for (int i = 0; i < n; i++) {

            int width = nse[i] - pse[i] - 1;

            int area = heights[i] * width;

            ans = Math.max(ans, area);
        }

        return ans;
    }
}